File: WalletApi/WalletApi/Models/Transaction.cs
namespace WalletApi.Models;

public class Transaction
{
    public Guid Id { get; private set; }
    public Guid WalletId { get; private set; }
    public TransactionType Type { get; private set; }
    public decimal Amount { get; private set; }
    public decimal BalanceAfter { get; private set; }
    public string? Description { get; private set; }
    public string? IdempotencyKey { get; private set; }
    public Guid? ReferenceTransactionId { get; private set; }
    public DateTime CreatedAtUtc { get; private set; }

    // Navigation
    public Wallet Wallet { get; private set; } = default!;

    private Transaction() { } // Required by EF Core

    public static Transaction Create(
        Guid walletId,
        TransactionType type,
        decimal amount,
        decimal balanceAfter,
        string? description = null,
        string? idempotencyKey = null,
        Guid? referenceTransactionId = null)
    {
        return new Transaction
        {
            Id = Guid.NewGuid(),
            WalletId = walletId,
            Type = type,
            Amount = amount,
            BalanceAfter = balanceAfter,
            Description = description,
            IdempotencyKey = idempotencyKey,
            ReferenceTransactionId = referenceTransactionId,
            CreatedAtUtc = DateTime.UtcNow
        };
    }
}

File: wallet_app/lib/features/transaction/domain/entities/transaction.dart
library;

import 'package:equatable/equatable.dart';

import 'package:wallet_app/features/transaction/domain/entities/transaction_type.dart';

class Transaction extends Equatable {
  final String id;
  final String walletId;
  final TransactionType type;
  final double amount;
  final double balanceAfter;
  final String? description;
  final String? idempotencyKey;
  final String? referenceTransactionId;
  final DateTime createdAtUtc;

  const Transaction({
    required this.id,
    required this.walletId,
    required this.type,
    required this.amount,
    required this.balanceAfter,
    this.description,
    this.idempotencyKey,
    this.referenceTransactionId,
    required this.createdAtUtc,
  });

  /// Whether this transaction increases the wallet balance.
  bool get isCredit =>
      type == TransactionType.deposit || type == TransactionType.transferIn;

  /// Whether this transaction decreases the wallet balance.
  bool get isDebit =>
      type == TransactionType.withdrawal || type == TransactionType.transferOut;

  @override
  List<Object?> get props => [
        id,
        walletId,
        type,
        amount,
        balanceAfter,
        description,
        idempotencyKey,
        referenceTransactionId,
        createdAtUtc,
      ];
}
